ruff_lsp/server.py

```diff
@@ -175,6 +204,149 @@ class VersionModified(NamedTuple):
+    version: int | None
+
+    @classmethod
+    def from_text_document(cls, text_document: workspace.TextDocument) -> Self:
+        """Create a `Document` from the given Text Document."""
+        return cls(
+            uri=text_document.uri,
+            path=text_document.path,
+            kind=DocumentKind.Text,
+            source=text_document.source,
+            version=text_document.version,
+        )
+
+    @classmethod
+    def from_notebook_document(cls, notebook_document: NotebookDocument) -> Self:
+        """Create a `Document` from the given Notebook Document."""
+        path = uris.to_fs_path(notebook_document.uri)
+        if path is None:
+            # `pygls` raises a `Exception` as well in `workspace.TextDocument`.
+            raise ValueError(
+                f"Unable to convert URI to file path: {notebook_document.uri}"
+            )
+        return cls(
+            uri=notebook_document.uri,
+            path=path,
+            kind=DocumentKind.Notebook,
+            source=_create_notebook_json(notebook_document),
+            version=notebook_document.version,
+        )
+
+    @classmethod
+    def from_uri(cls, uri: str) -> Self:
+        """Create a `Document` from the given URI.
+
+        The URI can be a file URI, a notebook URI, or a cell URI. The function will
+        try to get the notebook document first, and if that fails, it will fallback
+        to the text document.
+        """
+        notebook_document = LSP_SERVER.workspace.get_notebook_document(cell_uri=uri)
+        if notebook_document is None:
+            notebook_document = LSP_SERVER.workspace.get_notebook_document(
+                notebook_uri=uri
+            )
+        if notebook_document is None:
+            text_document = LSP_SERVER.workspace.get_text_document(uri)
+            return cls.from_text_document(text_document)
+        else:
+            return cls.from_notebook_document(notebook_document)
+
+    def is_stdlib_file(self) -> bool:
+        """Return True if the document belongs to standard library."""
+        return utils.is_stdlib_file(self.path)
+
+    def is_notebook_file(self) -> bool:
+        """Return True if the document belongs to a Notebook or a cell in a Notebook."""
+        return self.kind is DocumentKind.Notebook or self.path.endswith(".ipynb")
+
+
+SourceValue = Union[str, List[str]]
+
+
+class CodeCell(TypedDict):
+    """A code cell in a Jupyter notebook."""
+
+    cell_type: Literal["code"]
+    metadata: Any
+    outputs: list[Any]
+    source: SourceValue
+
+
+class MarkdownCell(TypedDict):
+    """A markdown cell in a Jupyter notebook."""
+
+    cell_type: Literal["markdown"]
+    metadata: Any
+    source: SourceValue
+
+
+class Notebook(TypedDict):
+    """The JSON representation of a Notebook Document."""
+
+    metadata: Any
+    nbformat: int
+    nbformat_minor: int
+    cells: list[CodeCell | MarkdownCell]
+
+
+def _create_notebook_json(notebook_document: NotebookDocument) -> str:
```

Comment:
Oh wow, this feels kind of expensive. Do we run this on every keystroke or only on save? How does it perform for large notebooks?

Reply:
Good questions. What about this function feels expensive? I guess the `json.dumps` part?

> Do we run this on every keystroke or only on save?

It depends on user settings but it could run on every keystroke when `"run": "onType"`.

> How does it perform for large notebooks?

I'll test it out.

We could potentially create our own layer of notebook synchronization but it feels not worth the effort if we're going to move the LSP to Rust.

Context: https://github.com/openlawlibrary/pygls/issues/394